File: LeetCode/C++/Hard/1639. Number of Ways to Form a Target String Given a Dictionary/1639-number-of-ways-to-form-a-target-string-given-a-dictionary.cpp
class Solution {
public:
    const int mod = 1e9+7;
    long dfs(vector<vector<long>>& dp, vector<vector<int>>& cnt, string& target, int i, int j, int n, int m) {
        if (j == m) return 1;
        if (n - i < m - j) return 0;
        if (dp[i][j] != -1) return dp[i][j];
        
        long val = cnt[i][target[j] - 'a'] * dfs(dp, cnt, target, i + 1, j + 1, n, m);
        val += dfs(dp, cnt, target, i + 1, j, n, m);
        val %= mod;
        return dp[i][j] = val;;
    }
    int numWays(vector<string>& words, string target) {
        int n = words[0].length();
        vector<vector<int>> cnt(n, vector<int>(26, 0));
        for (const auto& s: words) {
            for (int i = 0; i < n; i++) {
                cnt[i][s[i]-'a']++;
            }
        }
        
        int m = target.length();
        vector<vector<long>> dp(n, vector<long>(m, -1));
        return dfs(dp, cnt, target, 0, 0, n, m);
    }
};

File: LeetCode/Python3/Hard/1639. Number of Ways to Form a Target String Given a Dictionary/1639-number-of-ways-to-form-a-target-string-given-a-dictionary.py
class Solution:
    def numWays(self, words: List[str], target: str) -> int:
        m, n = len(words), len(words[0])
        l = len(target)
        d = [{} for i in range(n)]
        for i in range(m):
            for j in range(n):
                c = words[i][j]
                d[j][c] = d[j].get(c, 0) + 1
        @lru_cache(None)
        def dfs(i, j):
            if j == l:
                return 1
            elif n - i < l - j:
                return 0
            if target[j] not in d[i]:
                return dfs(i + 1, j)
            return (dfs(i + 1, j) + d[i][target[j]] * dfs(i + 1, j + 1)) % mod
        mod = 10 ** 9 + 7
        return dfs(0, 0)